submission/races.go

```diff
@@ -0,0 +1,215 @@
+			results <- result{sct: sct, logURL: logURL}
+		}(i, groupURLs[urlNum].Interface().(string))
+	}
+}
+
+func parallelNums(groups map[string]*ctpolicy.LogGroupInfo) map[string]int {
+	nums := make(map[string]int)
+	var subsetSum int
+	for _, g := range groups {
+		if !g.IsBase {
+			nums[g.Name] = g.MinInclusions
+			subsetSum += g.MinInclusions
+		}
+	}
+	nums[ctpolicy.BaseName] -= subsetSum
+	return nums
+}
+
+func totalLogs(groups map[string]*ctpolicy.LogGroupInfo) int {
+	unifiedLogs := make(map[string]bool)
+	for _, g := range groups {
+		if g.IsBase {
+			return len(g.LogURLs)
+		}
+		for l := range g.LogURLs {
+			unifiedLogs[l] = true
+		}
+	}
+	return len(unifiedLogs)
+}
+
+// AssignedSCT represents SCT with logURL of log-producer.
+type AssignedSCT struct {
+	sct    *ct.SignedCertificateTimestamp
+	logURL string
+}
+
+func groupsComplete(groups map[string]*ctpolicy.LogGroupInfo, sctList *guardedLogs, gCtx map[string]groupCtx) bool {
+	completeness := true
+	for groupName, g := range groups {
+		if sctList.sumFor(g.LogURLs) >= g.MinInclusions {
+			if gCtx[groupName].ctx.Err() != nil {
+				gCtx[groupName].cancel()
+			}
+		} else {
+			completeness = false
+		}
+	}
+	return completeness
+}
+
+type groupCtx struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func completenessError(groups map[string]*ctpolicy.LogGroupInfo, sctMap *guardedLogs, gCtx map[string]groupCtx) error {
+	cmplt := groupsComplete(groups, sctMap, gCtx)
+	if !cmplt {
+		return errors.New("Not all log-groups collected required number of SCTs")
+	}
+	return nil
+}
+
+// GetSCTs picks required number of Logs according to policy-group logic and collects SCTs from them.
+func GetSCTs(ctx context.Context, submitter Submitter, chain []ct.ASN1Cert, groups map[string]*ctpolicy.LogGroupInfo) ([]*AssignedSCT, error) {
+	logsNumber := totalLogs(groups)
+	results := make(chan result, logsNumber)
+	requestMap := guardedLogs{logURLs: make(map[string]bool)}
+	sctMap := guardedLogs{logURLs: make(map[string]bool)}
+	parallelNums := parallelNums(groups)
+	gCtx := make(map[string]groupCtx)
+	for _, g := range groups {
+		subCtx, cancel := context.WithCancel(ctx)
+		gCtx[g.Name] = groupCtx{ctx: subCtx, cancel: cancel}
+		defer gCtx[g.Name].cancel()
+		go func(group *ctpolicy.LogGroupInfo) {
+			groupRace(subCtx, chain, group, parallelNums[group.Name], results, &requestMap, &sctMap, submitter)
+		}(g)
+	}
+
+	var collectedSCTs []*AssignedSCT
+	var m sync.Mutex
+	for i := 0; i < logsNumber; i++ {
+		select {
+		case <-ctx.Done():
+			err := completenessError(groups, &sctMap, gCtx)
+			return collectedSCTs, err
+		case res := <-results:
+			if res.sct != nil {
+				m.Lock()
+				collectedSCTs = append(collectedSCTs, &AssignedSCT{sct: res.sct, logURL: res.logURL})
+				m.Unlock()
+				cmplt := groupsComplete(groups, &sctMap, gCtx)
+				if cmplt {
+					return collectedSCTs, nil
+				}
+			}
+		}
+	}
+	err := completenessError(groups, &sctMap, gCtx)
+	return collectedSCTs, err
```

Comment:
Function comment should mention that it's possible to have a result and a non-`nil` err.

Reply:
Done.